### ISSUE TITLE
[Issue Report]: Able to permanently kill player using cursed items

### ISSUE THREAD

**Issue by kphoenix137** [diasurgical/DevilutionX#6520]
### Operating System

Windows x64

### DevilutionX version

Other (please specify version number)

### Describe

https://github.com/diasurgical/devilutionX/assets/68359262/049c0821-e0af-4a55-9a03-25680b1b4e14

By equipping cursed items that bring your life below 1 and dying to a method that doesn't make you drop your items, restarting in town will result in a loop of your character appearing dead in town. The only way around this is to get another player to resurrect you.

### To Reproduce

1. equip cursed items
2. die without dropping items
3. restart in town

**Comment by kphoenix137** [comment]
Removing the call to `CalcPlrInv()` in `RestartTownLvl()` fixes this problem completely. No noticeable side effects. Is there a reason this function needs `CalcPlrInv()`?

**Comment by Trihedraf** [comment]
Wouldn't `CalcPlrInv()` be necessary for when you die and lose items?

**Comment by kphoenix137** [comment]
> Wouldn't `CalcPlrInv()` be necessary for when you die and lose items?

Yeah right here:
https://github.com/diasurgical/devilutionX/blob/45dbe6aa611810be103e14e503fb84b59915fa33/Source/player.cpp#L2739

**Comment by StephenCWills** [comment]
In Single Player, if the player chooses to save their game while their HP is negative, loading that savegame will crash the game client.

![image](https://github.com/user-attachments/assets/edd433f2-b4a2-4875-af9a-8e509d23fb45)

`InitPlayerGFX()` only loads the `player_graphic::Death` sprites, but the player wasn't dead when they saved the game, so the sprites needed to load the savegame don't get loaded.

https://github.com/diasurgical/devilutionX/blob/80ee3269111c46c8d340de57fedf6ce535826c43/Source/player.cpp#L2156-L2160

Furthermore, if the player equips an item that brings their HP below zero while in the dungeon, the game will instantly kill them. If the player does the same in town, the game prevents the player's death to avoid letting them die in town. If the player then enters the dungeon without removing the cursed items, the game does not instantly kill them. The game should probably consistently kill the player if the player's hp is negative while in the dungeon.

**Comment by kphoenix137** [comment]
I think there are 2 major issues that lead to undesirable gameplay:
1. Changing equipment can kill the player
2. Max life is allowed to go under 1

I think there is a potential solution to consider that might prevent the need for workarounds, such as the workaround that fixes town death via Healing/Heal Other, and whatever workaround would fix this issue report.

Prevent equipment changes that do either of the following:
- Reduces current life below 1
- Reduces max life below 1

**Comment by julealgon** [comment]
Why not just allow the player to die in town instead?

I feel like attempting to prevent equipment change based on the final effect of the change is both a convoluted solution, and very inflexible for further modding later.

For example, what if you want to have an item that drains 1hp every second that can kill the player, but you equip it with 2hp? Your check will not catch that because it wouldn't be an "instantaneous" death, but it would have a very similar effect regardless, so checking at all to me feels pointless.

**Comment by DakkJaniels** [comment]
> Why not just allow the player to die in town instead?
> 
I agree that checking for equipment change seems convoluted, but did you mean allow them to die in Town, or _not_ die in Town? If you allow dying in Town, and assuming equipment doesn't drop in Town, you'd never be able to be alive again.

I think just checking for player death when they enter the dungeon, as @StephenCWills said, makes the most sense and is the least complex solution.

**Comment by kphoenix137** [comment]
> Why not just allow the player to die in town instead?
> 
> I feel like attempting to prevent equipment change based on the final effect of the change is both a convoluted solution, and very inflexible for further modding later.
> 
> For example, what if you want to have an item that drains 1hp every second that can kill the player, but you equip it with 2hp? Your check will not catch that because it wouldn't be an "instantaneous" death, but it would have a very similar effect regardless, so checking at all to me feels pointless.

life drain =/= item bonus that changes life amount

**Comment by DakkJaniels** [comment]
I don't think those conclusions are necessary. Just make it so you can't die/be dead in Town, regardless of their max/current hp, and fix the issue with not dying in dungeon that @StephenCWills mentioned. If you want to provide some additional guardrails, you can do the "that would kill me" sound when someone tries to enter a TP or dungeon entrance from Town with max life <=0, but not strictly necessary.

**Comment by julealgon** [comment]
@kphoenix137 
> Life drain =/= item bonus that changes Life amount. Equipping a constricting ring would not be prevented with my suggestion at any life amount.

Fair enough. I wasn't trying to suggest they were the same thing.

> Also, you haven't suggested anything that fixes the problem outlined in this Issue report, but instead you seem to support the problem continuing to exist. ...  I'm not sure how you reasonably think the player should be permanently dead in town with no option to unbrick their character other than Resurrect.

Well, I did just post this right after:
> Drop the equipment in town and allow restarting, just like any normal death, is what I'd propose.

My point was just that I didn't see a reason to special-case death handling in town vs the dungeon.

And I know @StephenCWills brought up another scenario where death behavior is special-cased, but that doesn't diminish my point I don't think: I'd rather see _all_ special cases removed, if that was possible: if the player dies, he dies, drops the items where he died, and restarts on the specific tile in town. If the death happened in DLVL 16 or in town itself, I don't see a reason to change any of that behavior. 

Again, that's just my personal take. Apparently, most of you disagree with it which is fine.
 
> Then as far as changing Max Life to a negative number, this should not be allowed. It simply _doesn't make sense_. How can one have negative max life? The way it currently stands, it's possible to have a current Life greater than Max Life, and it's absurd. Example: 10/-50

Totally agree on this one. "Max life" should be capped at the bottom end like other attributes are/should. Bottom cap should probably be `1` for max life, and any reduction in max life instantly affects current life so that current life <= max life at all times.

> If you can think of any valid downsides to my solution, I am interested in hearing them so we can brainstorm for an improved solution.

The downside I see is not from a gameplay perspective, but from a game engine complexity perspective. When things work the same way as much as possible, the logic tends to become simpler to manage and less error prone. That's all.

Note as well that special cases will also tend to interact badly with modded content later, as mods will invariably introduce extra scenarios that you'd then want to special case as well because of the original customization. 

Imagine for example, a mod where there is a monster threat in town, and your character can now go into combat stance and attack/use spells in town to kill the monsters. With the extra "town-specific" workarounds you create here, you'll make realizing that mod harder as these restrictions would have to be lifted in several places in the code.

That is of course just an example.

**Comment by StephenCWills** [comment]
> And I know @StephenCWills brought up another scenario where death behavior is special-cased, but that doesn't diminish my point I don't think: I'd rather see _all_ special cases removed, if that was possible: if the player dies, he dies, drops the items where he died, and restarts on the specific tile in town. If the death happened in DLVL 16 or in town itself, I don't see a reason to change any of that behavior.

What does diminish your point is not the fact that it works differently, but that it works differently for a reason.

For PK, Blizzard chose to drop an ear as a trophy. I presume they didn't want gear to drop because it would be too much incentive for being a mean-spirited opportunist and diminish co-op play.

The reason dlvl 16 is a special case is because killing Diablo goes directly to the end movie and boots players from the game. There's too much opportunity to lose all your gear to a very poorly timed death.

Ignoring design decisions for the sake of reducing complexity is just short-sighted. Furthermore, and this is just my gripe, but your framing of "I don't see a reason to change any of that behavior" is a bit disingenuous considering you'd actually be the one changing behavior.

**Comment by StephenCWills** [comment]
Seems like something has reverted the fix. Try this...

1. Create a new Hellfire character
2. `dev.items.spawnUnique("warding")` and pick up the amulet
3. `dev.level.warp.dungeon(16)`
4. Equip the amulet

You will end up in an infinite death loop.

**Comment by kphoenix137** [comment]
Fixed by the following:
https://github.com/diasurgical/DevilutionX/pull/7939
https://github.com/diasurgical/DevilutionX/pull/8305
https://github.com/diasurgical/DevilutionX/pull/8302

Additional notes:
Somewhat relevant, but having negative life in town causes you to be unable to open up the ESC menu by pressing ESC, but still works by pressing the button in the panel.
